### PR TITLE
feat(injection): add app container name to envoy metadata

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR adds a new env var to the istio proxy container: `ISTIO_META_APP_CONTAINERS`. This env var is meant to capture the names of the application containers for which the sidecar proxy is being injected.

This is being added to improve telemetry reporting around the container to/from which traffic is flowing. Istio extensions can use the new var to have a direct link to the underlying containers. Stackdriver, for instance, can use this to select the appropriate `k8s_container` monitored resource. This replaces functionality lost across two different changes:
- the switch away from Mixer-based telemetry
- the dropping of the container port labelling requirement

This PR will not be considered to be complete until:
- [x] a proper run of `make gen` can be achieved

Gateway proxies will not have values for `ISTIO_META_APP_CONTAINERS`, as there is no such thing for gateways.

[ X ] Policies and Telemetry

### CODE REVIEW

File: manifests/istio-control/istio-discovery/files/injection-template.yaml [hunk 1]
```diff
@@ -191,6 +191,17 @@ template: |
           {{- end}}
         {{- end}}
         ]
+    - name: ISTIO_META_APP_CONTAINER
```
Comment (bianpengyuan) [inline manifests/istio-control/istio-discovery/files/injection-template.yaml:194]: Probably better to just list all containers? Not the sd use case but there might be extensions that need to look up all containers in the pod. Consolidating multiple app containers and only proxy container into `istio-proxy` is also a bit of confusing and might not uniformly apply to all extensions. It should be extension logic to decide how to properly consume container information.

Comment (mandarjog) [reply manifests/istio-control/istio-discovery/files/injection-template.yaml:194]: What. Considerations can extensions apply to pick one name from another?
I suppose this is saying designate the most important one.
We could make it a list with earlier in the list means more important.
And then the extension can take additional config.
Making it a list feels like Yagni.

**Comment by howardjohn** [comment]
As mentioned offline, we fundamentally cannot determine which container traffic is coming from in kubernetes. Do we really want to report a metric that is a "guess"? Even a single container we can still be wrong, as it could come from istio-proxy, for example

File: manifests/istio-control/istio-discovery/files/injection-template.yaml [hunk 2]
```diff
@@ -191,6 +191,17 @@ template: |
...
+      {{- if eq (len .Spec.Containers) 1 }}
+        {{- $first := true }}
+        {{- range $index, $container := .Spec.Containers }}
```
Comment (howardjohn) [inline manifests/istio-control/istio-discovery/files/injection-template.yaml:197]: we can probably simplify this with the `first` function

**Comment by mandarjog** [comment]
We all understand @howardjohn that it is not possible to know which container the traffic came from. This work is for destination container, not source.

**Comment by mandarjog** [comment]
@douglas-reid with one other container it is clear that we want to associate traffic this it. With more that one it will be like today where we associate everything with istio-proxy which is what we do today. But we know that these metrics are *not* for istio-proxy so we are choosing a container that we surely know if incorrect.

I am wondering if we can use “unknown”,  will sd like a container that does not exist at all? Will the linking from metrics to physical resources work? Otherwise istio-proxy is the representative of the pod that does not break linkage.

**Comment by howardjohn** [comment]
How do you know which container is the destination? If I get inbound traffic for port 80, it could go to either container, right?

**Comment by bianpengyuan** [comment]
The metrics that we are collecting is application request metrics, so target container should be app container instead of proxy. Using proxy as target container is just a compromise. Under what circumstance does traffic targets istio-proxy instead of application container? Health check prober is the only thing I can think of. We should filter out those requests from application request metrics, instead of living with it and delivering less than ideal user experience.

**Comment by mandarjog** [comment]
Istio-proxy container is an implementation detail of the service mesh.
A user actually cares about which application container the traffic is going to.
It is very rare that traffic was sent to an IP address with the intent of sending to the proxy. Client always wants to send it to the application. Istio-proxy is a compromise since in the data plane we only know the existence of that container. It is also the exactly the wrong thing :-( so we want to do something better.

**Comment by howardjohn** [comment]
>  a proper run of make gen can be achieved (broken on my linux desktop)

Can you file an issue?

**Comment by douglas-reid** [comment]
Thanks for the great reviews!

I decided to adopt @bianpengyuan 's suggestion of just returning a list of application containers. This moves the decision of how/when to use that information to the extensions themselves.

I think we can then have a discussion of how any particular extension uses the information in the extension-specific PR (examples: should SD use `unknown` instead of `istio-proxy`, when we should assume traffic is to the proxy itself, etc.).

**Comment by douglas-reid** [comment]
Gentle ping for review.

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
@douglas-reid I put a hold on this since this is for telemetry and I cannot really properly review in that context. But if you don't want review from someone else in telemetry feel free to remove it, LGTM

**Review by bianpengyuan — APPROVED** [review]
thanks!

**Pull request merged**